### DISCUSSION TITLE
Extended configuration options

### DISCUSSION THREAD

**Discussion by zltn** [filamentphp/filament#58] [Ideas]
Just installed this for a new projects and super happy how quickly I got it up and running. A couple of configuration options I miss or couldn't find:

- [x] Re-use my existing user table vs creating a new one, believe there's already a thread for that [here](https://github.com/laravel-filament/filament/discussions/30). This would be by far the most important one - Fixed with #72 
- [x] Configure the top widgets on the dashboard. The only way I found was to publish the views (which is also not documented) - #71 fixed this
- [x] Sort the resources in the side menu found the `$navigationSort` property on the resource but does not seem to respond to any number.~~ Bug fixed with [this commit](https://github.com/laravel-filament/filament/commit/35c3751acf5439fb9fd011dca44943ca2cf09620)
- [x] be able to de-activate edit and create for some resources, ideally even per user(type) - added in 1.5
- [x] Configure colors. I tried publishing the assets but the css is already minified. Would be great if we could use tailwind configurations for this - added in 1.5
- [ ] Edit resource behavior after saving. Right now you return to the same page, but there's no back button. Would be great to be able to configure to enter go back to the table or keep existing behavior. For the existing behavior I'd recommend adding a clear back button to the table view

**Comment by danharrin** [comment]
Hey @zltn. I've opened #72 with support for custom user tables. Would be great if you could review, thanks!

**Reply by danharrin** [reply]
This is now supported from v1.3.0.

**Reply by zltn** [reply]
@danharrin just implemented and noticed a minor security issue in the isFilamentUser trait:

```php

    public function canAccessFilament()
    {
        $column = static::getFilamentUserColumn();

        return $column !== null ?
            $this->{$column} :
            true;
    }

    public function isFilamentAdmin()
    {
        $column = static::getFilamentAdminColumn();

        return $column !== null ?
            $this->{$column} :
            true;
    }
```

It would probably be better to invert these and make the default return false. If for what ever reason something upstream fails everyone in the user table would have access. I found it by logging in before running the migration and was surprised I got access :) 

I didn't check the roles logic but might apply to that as well

**Reply by zltn** [reply]
Another small point. How should we handle the default filament user migration? Maybe make the filament user migration optional (publish it first if you want to use a the default )

**Reply by danharrin** [reply]
Hey @zltn, I appreciate the concern but I don’t think this is a security issue? The default behaviour ensures that custom setups work out of the box, and we encourage users to override these methods to build their own setup anyway. We won’t be making the users migration optional, that would just be an extra installation step that 95% of users would not benefit from. If you’d like to remove the tables from your app, simply write a migration that deletes them. Thanks!

**Reply by zltn** [reply]
Agree on the migration, that's what I did for my use case, just wasn't sure which one of the two methods is (will be) more popular.

For the two methods above I'd strongly encourage to flip the check and only return true if the column is set and the value is true. Returning false in all other cases will still make it work out of the box but no user will be authorized by default which I'd consider safer vs making all users filament-users and admins by default by just adding the trait. Laravel's gates work similarly where the default return is void and only return true if the statement is explicitly true

**Reply by danharrin** [reply]
After discussing this with @ryangjchandler, we agree that this approach is safer, and have removed admin permissions from users by default. They will still be able to access Filament, but this feels like a good compromise between ease of use and maximum security. [The docs now clearly detail how to restrict access to Filament](https://filamentadmin.com/docs/#users), and the recommended approach will remain with the default `filament_users` table.

**Reply by zltn** [reply]
Awesome, looks like the new docs haven't been deployed yet but read through here and looks good https://github.com/laravel-filament/docs/commit/cf56887d32499ab110d34ff388c03b8b58c9b234

**Reply by danharrin** [reply]
Yeah we were having a few issues getting them to build. All fixed now, thanks!

**Comment by danharrin** [comment]
Nav sorting is fixed in 35c3751, soon to be released.

**Reply by danharrin** [reply]
Supported from v1.3.0.

**Comment by danharrin** [comment]
The widgets will be configurable when #71 is reviewed.

**Reply by danharrin** [reply]
Also supported from v1.3.0.

**Comment by danharrin** [comment]
The primary colour will be configurable in the near future through a configuration option.

**Comment by danharrin** [comment]
You may customise save behaviour by overriding the `save()` method on the edit page class. We may provide an option for "Save", "Save & Close" etc in the future, would that fit?

**Reply by zltn** [reply]
yeah, making the action configurable would even be great. Even better if it's based on roles/policies. One more example I ran into was me by accident deleting a role vs just detaching it. 

See the screenshot below, in this case I'd def would like to disable the delete button before someone accidentally deletes a role and the entire system breaks. 

<img width="592" alt="Screen Shot 2021-03-09 at 9 20 59 PM" src="https://user-images.githubusercontent.com/1194401/110565995-5b41b680-811d-11eb-89b9-d3db73957e43.png">

**Reply by danharrin** [reply]
#106 supports this :)

**Reply by zltn** [reply]
Awesome, will test it out once  it merges

**Reply by zltn** [reply]
@danharrin just updated and running into two things that I couldn't find in the docs:

- this assumes you're using Filament's default User model + roles. I'm using a custom User model + Spatie's permission for roles. I saw there was a `getFilamentRolesColumn` stub so should I assume that I have to store the roles only in a column on the user model vs a separate table like Spatie does ? Would also be good to document the roles column behavior in the docs when using a custom User model

- I saw the `authorization()` stub on the resource but couldn't figure out how this would work on a per model base and how this would work for relationships like my example above

**Reply by zltn** [reply]
looks like I was a bit to eager, only merged in the dev branch, not main. But stil curious how the roles would work if they are not stored in the user table

**Reply by danharrin** [reply]
@zltn roles are an option only to apps that choose to have a JSON `roles` column, specifically for Filament, on their users table. Everything that can be done with roles can be done with policies, they just keep auth logic together with the resource and save the extra class 👍

**Reply by ryangjchandler** [reply]
@zltn To use your own permissions, I'd recommend creating a policy for your models:

```php
class OrderPolicy
{
    use HandlesAuthorization;

    public function create(User $user, Order $order)
    {
        return $user->hasRole('name-of-role');
    }
}
```

**Reply by zltn** [reply]
perfect, just tested with the dev branch and works

**Comment by danharrin** [comment]
We will have support for Laravel authorisation policies soon, which will allow you to restrict edit and create access.

**Reply by sbrow** [reply]
The current documentation states that filament observes model policies. Is this currently true, or is it still in the works? I can't get it to work for me.
![image](https://user-images.githubusercontent.com/6729162/114923687-1fc09900-9dfb-11eb-96d5-5e35d9dd31b0.png)

**Reply by danharrin** [reply]
@sbrow they should work, what have you tried that doesn't behave correctly?

**Reply by YunshuaiPeng** [reply]
Same confused. My UserPolicy delete method return false. But I can delete user still.

**Reply by danharrin** [reply]
@pys1992 is the policy registered in Laravel?

**Reply by YunshuaiPeng** [reply]
@danharrin Yes. It registered by the [auto discovery behavior](https://laravel.com/docs/8.x/authorization#policy-auto-discovery).
I did these steps:
1. `php artisan make:policy UserPolicy --model=User` create user policy.
2. Alter the delete method to return false.
3. `php artisan make:filament-resource User` create user resource.

Then I can see users on admin page. Finally I select a user and click delete button, user deleted.

Did I miss any steps?

**Reply by danharrin** [reply]
Ah, @pys1992 if you're using an admin account then you override all permissions automatically. Create a new user account without superuser perms.
